pallets/parachain-staking/src/lib.rs

```diff
@@ -525,7 +525,7 @@ pub mod pallet {
 				new_total,
 			);
 
-			10_000
+			300_000_000_000 // Three fifths of the max block weight
```

Comment:
This may be an overestimate. I would assume the risk of an overestimate is that if the block for which the runtime upgrade is scheduled has 2/5th of max block weight in other extrinsics, then setting this value to 3/5ths of max block weight would lead to the relay chain not accepting/finalizing the block (even if the actual computation was less than this weight value).

What are the risks of underestimating weight for this computation? Right now, all runtime methods in staking (except `set_inflation`) underestimate weight including `on_finalize` every Round.

Reply:
I think an ideal solution would be to query the currently consumed block weight and keep up with how much we're spending as we iterate (using benchmarking for estimates), then return when we can't afford another iteration. If we end up writing any sort of infrastructure for these runtime upgrades, that could be worthwhile.

Reply:
The accurate weight estimate is `num_of_collators * COST_OF_READ + 1 * COST_OF_READ + 1 * COST_OF_WRITE`, because there is 1 read, 1 write for `Total` and `num_of_collators` reads of `CollatorState`. There is also the computation associated with computing `new_total`.

I'll bring back benchmarking in another PR today/tomorrow. It'll be useful for making estimates, but I still have the questions:
1. what is the risk of overestimating weight?
2. what is the risk of underestimating weight?

I think (1) >> (2) when it comes to stalling a live chain. I'm unaware of the risks of underreporting weights in this context.

Reply:
* If we do more work than we charge for, we have security problems / risk of stalling a chain.
* If we do less work than we charge for, we just don't get as much done as we could.
* If we somehow end up with more weight in a block than is allowed, it's invalid.

Reply:
> If we do less work than we charge for, we just don't get as much done as we could.

If we overcharge for computation and it leads to (3), then we stall the chain because the block with the runtime upgrade will be invalid.

> If we do more work than we charge for, we have security problems / risk of stalling a chain.

If we underestimate weight consistently, I agree there are security problems, but I don't see how it stalls the chain unless it's an excessive amount of computation.

This is a one-off computation so I still have a feeling that underestimating is safer than overestimating.

Reply:
Agreed. When we can no longer fit a migration within one block, though, the weights become a lot more important.

Reply:
> if the block for which the runtime upgrade is scheduled has >2/5th of max block weight in other extrinsics, then overestimating at 3/5ths of max block weight would lead to the relay chain not accepting/finalizing the block

Yes, that's right. The relay chain would reject a block that had this migration followed by that many extrinsics.

The reason that overestimating doesn't lead to this situation is that a un-hacked authoring node would not construct such a block. FRAME's block construction process is to apply on_runtime_upgrade and on_initialize first and thus count their weight first. (This is why on_initialize is supposed to account for the weight of on_finalize as well.) Then start including extrinsics making sure that there is space left before each extrinsic.